File: src/main/java/com/github/sorusclient/client/module/impl/perspective/Perspective.java
package com.github.sorusclient.client.module.impl.perspective;

import com.github.glassmc.loader.GlassLoader;
import com.github.sorusclient.client.Sorus;
import com.github.sorusclient.client.adapter.*;
import com.github.sorusclient.client.event.EventManager;
import com.github.sorusclient.client.adapter.event.KeyEvent;
import com.github.sorusclient.client.module.ModuleDisableable;
import com.github.sorusclient.client.setting.Setting;
import com.github.sorusclient.client.setting.SettingConfigurableData;

import java.util.List;

public class Perspective extends ModuleDisableable {

    private final Setting<Key> key;

    private boolean toggled;
    private PerspectiveMode previousPerspective = null;

    public Perspective() {
        super("perspective");

        this.register("key", this.key = new Setting<>(Key.F));

        Sorus.getInstance().get(EventManager.class).register(KeyEvent.class, this::onKey);
    }

    private void onKey(KeyEvent event) {
        IAdapter adapter = Sorus.getInstance().get(IAdapter.class);
        if (this.isEnabled() && adapter.getOpenScreen() == ScreenType.IN_GAME) {
            if (event.getKey() == this.key.getValue() && !event.isRepeat()) {
                this.toggled = event.isPressed();

                if (this.toggled) {
                    this.previousPerspective = adapter.getPerspective();

                    GlassLoader.getInstance().getInterface(IPerspectiveHelper.class).onToggle();

                    adapter.setPerspective(PerspectiveMode.THIRD_PERSON_BACK);
                } else {
                    adapter.setPerspective(this.previousPerspective);
                }
            }
        }
    }

    public boolean isToggled() {
        return toggled;
    }

    @Override
    public void addSettings(List<SettingConfigurableData> settings) {
        super.addSettings(settings);
        settings.add(new SettingConfigurableData("Key", this.key, SettingConfigurableData.ConfigurableType.KEYBIND));
    }

}

File: src/main/kotlin/com/github/sorusclient/client/module/impl/perspective/Perspective.kt
package com.github.sorusclient.client.module.impl.perspective

import com.github.glassmc.loader.GlassLoader
import com.github.sorusclient.client.adapter.AdapterManager
import com.github.sorusclient.client.adapter.Key
import com.github.sorusclient.client.adapter.PerspectiveMode
import com.github.sorusclient.client.adapter.ScreenType
import com.github.sorusclient.client.adapter.event.KeyEvent
import com.github.sorusclient.client.event.EventManager
import com.github.sorusclient.client.module.ModuleDisableable
import com.github.sorusclient.client.setting.ConfigurableData
import com.github.sorusclient.client.setting.ConfigurableData.KeyBind
import com.github.sorusclient.client.setting.Setting

class Perspective : ModuleDisableable("perspective") {

    private var key: Setting<Key>
    var isToggled = false
        private set
    private var previousPerspective: PerspectiveMode = PerspectiveMode.FIRST_PERSON

    init {
        register("key", Setting(Key.F).also { key = it })
        EventManager.register(this::onKey)
    }

    private fun onKey(event: KeyEvent) {
        val adapter = AdapterManager.getAdapter()
        if (isEnabled() && adapter.openScreen == ScreenType.IN_GAME) {
            if (event.key == key.value && !event.isRepeat) {
                isToggled = event.isPressed
                if (isToggled) {
                    previousPerspective = adapter.perspective
                    GlassLoader.getInstance().getInterface(IPerspectiveHelper::class.java).onToggle()
                    adapter.perspective = PerspectiveMode.THIRD_PERSON_BACK
                } else {
                    adapter.perspective = previousPerspective
                }
            }
        }
    }

    override fun addSettings(settings: MutableList<ConfigurableData>) {
        super.addSettings(settings)
        settings.add(KeyBind("Key", key))
    }

}